File: GanServer/src/com/xattacker/binary/TypeConverter.java
package com.xattacker.binary;

public class TypeConverter
{
	// in order to make constructor hide
	protected TypeConverter()
	{
	}

	public static byte[] charToByte(char aChar)
	{
		byte[] bytes = new byte[2];
		bytes[0] = (byte) ((aChar & 0xff00) >> 8);  
		bytes[1] = (byte) (aChar & 0x00ff);   

		return bytes;
	}
	
	public static byte[] charArrayToByteArray(char[] aChars)
	{
		byte[] bytes = new byte[aChars.length*2]; 
		
		for (int i = 0, length = aChars.length; i < length; i++) 
		{    
			bytes[2*i] = (byte) ((aChars[i] & 0xff00) >> 8);    
			bytes[2*i+1] = (byte) (aChars[i] & 0x00ff); 
		}
		
		return bytes;
	}

	public static char byteToChar(byte[] aBytes)
	{
		int value = aBytes[0];
		
		value &= 0xff;
		value |= ((short) aBytes[1] << 8);

		return (char) value;
	}

	public static byte[] shortToByte(short aShort)
	{
		short temp = aShort;
		byte[] bytes = new byte[2];

		for (int i = 0; i < bytes.length; i++)
		{
			bytes[i] = new Integer(temp & 0xff).byteValue();
			temp = (short)(temp >> 8);
		}

		return bytes;
	}

	public static short byteToShort(byte[] aBytes)
	{
		short value = (short) aBytes[0];
		
		value &= 0xff;
		value |= ((short) aBytes[1] << 8);

		return value;
	}
	
	public static byte[] intToByte(int aInt)
	{
		int temp = aInt;
		byte[] bytes = new byte[4];

		for (int i = 0; i < bytes.length; i++)
		{
			bytes[i] = new Integer(temp & 0xff).byteValue();
			temp = temp >> 8;
		}

		return bytes;
	}

	public static int byteToInt(byte[] aBytes)
	{
		int value = aBytes[0];
		
		value &= 0xff;
		value |= (aBytes[1] << 8);
		
		value &= 0xffff;
		value |= (aBytes[2] << 16);
		
		value &= 0xffffff;
		value |= (aBytes[3] << 24);

		return value;
	}

	public static byte[] longToByte(long aLong)
	{
		long temp = aLong;
		byte[] bytes = new byte[8];
		
		for (int i = 0; i < bytes.length; i++)
		{
			bytes[i] = new Integer((int)temp & 0xff).byteValue();
			temp = temp >> 8;
		}

		return bytes;
	}

	public static long byteToLong(byte[] aBytes)
	{
		long value = aBytes[0];
		
		value &= 0xff;
		value |= ((long) aBytes[1] << 8);
		
		value &= 0xffff;
		value |= ((long) aBytes[2] << 16);
		
		value &= 0xffffff;
		value |= ((long) aBytes[3] << 24);
		
		value &= 0xffffffffl;
		value |= ((long) aBytes[4] << 32);
		
		value &= 0xffffffffffl;
		value |= ((long) aBytes[5] << 40);
		
		value &= 0xffffffffffffl;
		value |= ((long) aBytes[6] << 48);
		
		value &= 0xffffffffffffffl;
		value |= ((long) aBytes[7] << 56);
		
		return value;
	}
	
	public static byte[] doubleToByte(double aDouble)
	{
		return longToByte(Double.doubleToLongBits(aDouble));
	}
	
	public static double byteToDouble(byte[] aBytes)
	{
		return Double.longBitsToDouble(byteToLong(aBytes));
	}
}

File: GanAndroidApp/GanClient/src/main/java/com/xattacker/binary/TypeConverter.kt
package com.xattacker.binary

import android.app.Activity
import kotlin.experimental.and
import kotlin.experimental.or

object TypeConverter
{
    const val CHAR_SIZE = 2
    const val SHORT_SIZE = 2
    const val INTEGER_SIZE = 4
    const val LONG_SIZE = 8
    const val DOUBLE_SIZE = 8

    fun charToByte(aChar: Char): ByteArray
    {
        val bytes = ByteArray(CHAR_SIZE)
        bytes[0] = (aChar.toInt() and 0xff00 shr 8).toByte()
        bytes[1] = (aChar.toInt() and 0x00ff).toByte()

        return bytes
    }

    fun charArrayToByte(aChars: CharArray): ByteArray
    {
        val bytes = ByteArray(aChars.size * 2)

        var i = 0
        val length = aChars.size
        while (i < length)
        {
            bytes[2 * i] = (aChars[i].toInt() and 0xff00 shr 8).toByte()
            bytes[2 * i + 1] = (aChars[i].toInt() and 0x00ff).toByte()
            i++
        }

        return bytes
    }

    fun byteToChar(aBytes: ByteArray): Char
    {
        var value = aBytes[0].toInt()

        value = value and 0xff
        value = value or (aBytes[1].toInt() shl 8)

        return value.toChar()
    }

    fun shortToByte(aShort: Short): ByteArray
    {
        var temp = aShort
        val bytes = ByteArray(SHORT_SIZE)

        for (i in bytes.indices)
        {
            val b = Integer.valueOf(temp.toInt() and 0xff)
            bytes[i] = b.toByte()
            temp = (temp.toInt() shr 8).toShort()
        }

        return bytes
    }

    fun byteToShort(aBytes: ByteArray): Short
    {
        var value = aBytes[0].toShort()

        value = value and 0xff
        value = value or (aBytes[1].toInt() shl 8).toShort()

        return value
    }

    fun intToByte(aInt: Int): ByteArray
    {
        var temp = aInt
        val bytes = ByteArray(INTEGER_SIZE)

        for (i in bytes.indices)
        {
            val b = Integer.valueOf(temp and 0xff)
            bytes[i] = b.toByte()

            temp = temp shr 8
        }

        return bytes
    }

    fun byteToInt(aBytes: ByteArray): Int
    {
        var value = aBytes[0].toInt()

        value = value and 0xff
        value = value or (aBytes[1].toInt() shl 8)

        value = value and 0xffff
        value = value or (aBytes[2].toInt() shl 16)

        value = value and 0xffffff
        value = value or (aBytes[3].toInt() shl 24)

        return value
    }

    fun longToByte(aLong: Long): ByteArray
    {
        var temp = aLong
        val bytes = ByteArray(LONG_SIZE)

        for (i in bytes.indices)
        {
            val b = Integer.valueOf(temp.toInt() and 0xff)
            bytes[i] = b.toByte()

            temp = temp shr 8
        }

        return bytes
    }

    fun byteToLong(aBytes: ByteArray): Long
    {
        var value = aBytes[0].toLong()

        value = value and 0xff
        value = value or (aBytes[1].toLong() shl 8)

        value = value and 0xffff
        value = value or (aBytes[2].toLong() shl 16)

        value = value and 0xffffff
        value = value or (aBytes[3].toLong() shl 24)

        value = value and 0xffffffffL
        value = value or (aBytes[4].toLong() shl 32)

        value = value and 0xffffffffffL
        value = value or (aBytes[5].toLong() shl 40)

        value = value and 0xffffffffffffL
        value = value or (aBytes[6].toLong() shl 48)

        value = value and 0xffffffffffffffL
        value = value or (aBytes[7].toLong() shl 56)

        return value
    }

    fun doubleToByte(aDouble: Double): ByteArray
    {
        return longToByte(java.lang.Double.doubleToLongBits(aDouble))
    }

    fun byteToDouble(aBytes: ByteArray): Double
    {
        return java.lang.Double.longBitsToDouble(byteToLong(aBytes))
    }
}
